Mathlib/Topology/CWComplex/Classical/Basic.lean

```diff
@@ -523,38 +523,220 @@ lemma CWComplex.cellFrontier_subset_finite_openCell [CWComplex C] (n : ℕ) (i :
     cellFrontier n i ⊆ ⋃ (m < n) (j ∈ I m), openCell m j := by
   simpa using RelCWComplex.cellFrontier_subset_finite_openCell n i
 
+section Subcomplex
+
+namespace RelCWComplex
+
+/-- A subcomplex is a closed subspace of a CW complex that is the union of open cells of the
+  CW complex. -/
+structure Subcomplex (C : Set X) {D : Set X} [RelCWComplex C D] where
+  /-- The underlying set of the subcomplex. -/
+  carrier : Set X
+  /-- The indexing set of cells of the subcomplex. -/
+  I : Π n, Set (cell C n)
+  /-- A subcomplex is closed. -/
+  closed' : IsClosed carrier
+  /-- The union of all open cells of the subcomplex equals the subcomplex. -/
+  union' : D ∪ ⋃ (n : ℕ) (j : I n), openCell (C := C) n j = carrier
+
+namespace Subcomplex
+
+variable [RelCWComplex C D]
+
+instance : SetLike (Subcomplex C) X where
+  coe E := E.carrier
+  coe_injective' E F h := by
+    obtain ⟨E, _, _, hE⟩ := E
+    obtain ⟨F, _, _, hF⟩ := F
+    congr
+    apply eq_of_eq_union_iUnion
+    rw [hE, hF]
+    simpa using h
+
+lemma mem_carrier {E : Subcomplex C} {x : X} : x ∈ E.carrier ↔ x ∈ (E : Set X) := Iff.rfl
+
+lemma coe_eq_carrier {E : Subcomplex C} : (E : Set X) = E.carrier := rfl
+
+lemma mem {E : Subcomplex C} {x : X} : x ∈ E ↔ x ∈ (E : Set X) := Iff.rfl
+
+@[ext] lemma ext {E F : Subcomplex C} (h : ∀ x, x ∈ E ↔ x ∈ F) : E = F :=
+  SetLike.ext h
+
+lemma eq_iff (E F : Subcomplex C) : E = F ↔ (E : Set X) = F :=
+  SetLike.coe_injective.eq_iff.symm
+
+/-- Copy of a `Subcomplex` with a new `carrier` equal to the old one. Useful to fix definitional
+equalities. See Note [range copy pattern]. -/
+protected def copy (E : Subcomplex C) (F : Set X) (hF : F = E) : Subcomplex C :=
+  { carrier := F
+    I := E.I
+    closed' := hF.symm ▸ E.closed'
+    union' := hF.symm ▸ E.union' }
+
+@[simp] lemma coe_copy (E : Subcomplex C) (F : Set X) (hF : F = E) :
+  (E.copy F hF : Set X) = F := rfl
+
+lemma copy_eq (E : Subcomplex C) (F : Set X) (hF : F = E) : E.copy F hF = E :=
+  SetLike.coe_injective hF
+
+lemma union (E : Subcomplex C) :
+    D ∪ ⋃ (n : ℕ) (j : E.I n), openCell (C := C) (D := D) n j.1 = E := by
+  rw [E.union']
+  rfl
+
+lemma closed (E : Subcomplex C) : IsClosed (E : Set X) := E.closed'
+
+end Subcomplex
+
+end RelCWComplex
+
+namespace CWComplex
+
+export RelCWComplex (Subcomplex)
+
+namespace Subcomplex
+
+export RelCWComplex.Subcomplex (I closed union mem_carrier coe_eq_carrier mem ext copy coe_copy
+  copy_eq)
+
+end CWComplex.Subcomplex
+
+lemma CWComplex.Subcomplex.union {C : Set X} [CWComplex C] {E : Subcomplex C} :
+    ⋃ (n : ℕ) (j : E.I n), openCell (C := C) n j = E := by
+  have := RelCWComplex.Subcomplex.union E (C := C) (D := ∅)
+  rw [empty_union] at this
+  exact this
+
+/-- An alternative version of `Subcomplex`: Instead of requiring that `E` is closed it requires
+  that for every cell of the subcomplex the corresponding closed cell is a subset of `E`. -/
```

Comment:
```suggestion
/-- An alternative version of `Subcomplex.mk`: Instead of requiring that `E` is closed it requires
  that for every cell of the subcomplex the corresponding closed cell is a subset of `E`. -/
```